[MRG] Allow kwarg handles and labels figure.legend and make doc for kwargs the same

jklymak opened the pull request:
<!--Thank you so much for your PR! To help us review, fill out the form
to the best of your ability.  Please make use of the development guide at
https://matplotlib.org/devdocs/devel/index.html-->

<!--Provide a general summary of your changes in the title above, for
example "Raises ValueError on Non-Numeric Input to set_xlim".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

<!--If you are able to do so, please do not create the
PR out of master, but out of a separate branch.  See
https://matplotlib.org/devel/gitwash/development_workflow.html for
instructions.-->

## PR Summary

Allows `fig.legend(handles=hs, labels=labs)` keyword calls.  

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

Fixes #9320

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [x] Check reworked docs are correct... 

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

dstansby commented:
Thanks for fixing this. I can't think of any reason not to just have them as kwargs with defaults of `None` as you said in the bug report, but maybe I'm missing something.

tacaswell commented on lib/matplotlib/figure.py:
```diff
@@ -1461,8 +1461,41 @@ def legend(self, *args, **kwargs):
+                    warnings.warn('The handle {!r} has a label of {!r} which '
+                                  'cannot be automatically added to the '
+                                  'legend.'.format(handle, label))
+                    labels.remove(label)
```
Can we let this be handled down stream?

jklymak replied:
The underscore check?  Maybe.  I just shamelessly copied from axes.legend. 

Yes yes I know I should have refactored it... otoh I don’t like that axes.legend if there is only one argument assumes that it is a list of labels.  I think a list of handles makes much more sense.

jklymak replied:
Put this check into `legend.Legend.__init__`, and removed from `axes.legend` and `figure.legend`

tacaswell replied:
Taking labels (not handles) is a bit problematic and leads to issues like https://github.com/benlaken/European_wind/pull/1

jklymak replied:
@tacaswell. Um yeah that does look to be a bad decision. 

I’d suggest we just strip the illegal underscore and warn (or logging.warning 😉) or set it to an empty string.  

But I didn’t originate this code.  I’m not actually sure why an underscore is a problem in the first place.

tacaswell commented on lib/matplotlib/tests/test_legend.py:
```diff
@@ -74,6 +74,56 @@ def test_labels_first():
     ax.legend(loc=0, markerfirst=False)
 
 
+@image_comparison(baseline_images=['legend_fig_noargs'], extensions=['png'],
```
Is there a reasonable way to check that this works without an image test?

jklymak replied:
Happy to try that.  Or reduce the tests.  I don’t know how you would test it without a plot but that’s just my ignorance.  Happy to implement something

jnothman replied:
You only need to test equivalence between the results of different invocations, not equivalence to some human-verified gold standard output.

anntzer replied:
see https://github.com/matplotlib/matplotlib/pull/9305/files#diff-663c17d0729218d60dbb862c286a17d4R5327

jklymak replied:
OK< I think I understand.  Use the trick linked above and then assert that the pngs are the same w/o actually having to save them or do the image comparison?  That sounds great.  

Sorry for being slow, but I don't know *anything* about writing tests for python...

jnothman replied:
And I'm not an expert at testing with matplotlib, but I would've thought there might be a way to check the calculated legend artists are equivalent. But perhaps not.

tacaswell replied:
`ax.legend_.get_texts()` and `ax.legend_.legendHandles`  _seems_ to get you back the labels and handles used, but that is just from poking at a legend object interactively....

jklymak commented:
@dstansby  Yes, I'm still not sure.  But I largely copied from `axes.legend` and its done with `*args, *kwargs` as well, so I did that here.

jklymak commented:
OK, I refactored.  Not `figure.legend` and `axes.legend` have exactly the same argument parsing language (now in `legend._parse_legend_args`).  

Probably need the docs for `figure.legend` changed.  Its possible this is non-backwards compatible.

This is not passing a legend test that I don't understand so I'm having difficulty debugging...

```python

def test_legend_handler_map(self):
        lines = plt.plot(range(10), label='hello world')
        with mock.patch('matplotlib.axes.Axes.'
                        'get_legend_handles_labels') as handles_labels:
            handles_labels.return_value = lines, ['hello world']
            plt.legend(handler_map={'1': 2})
        handles_labels.assert_called_with({'1': 2})
```

jklymak commented:
It seems that both instances of legend should have pass-through kwarg documentation.

It seems the *right* way to do this is for each keyword argument to have a `set_arg` that has an `ACCEPTS:` line in the doc.  But I'm not doing that.  So I assume I can define a kw doc string and then reference it somehow.

tacaswell commented:
Well, that escalated quickly.  I thought this was going to be a ~5 line fix....

anntzer commented on lib/matplotlib/legend.py:
```diff
@@ -105,6 +106,177 @@ def _update_bbox_to_anchor(self, loc_in_canvas):
         self.legend.set_bbox_to_anchor(loc_in_bbox)
 
 
+legend_kw_doc = '''
```
make this `_private`?

jklymak commented:
@tacaswell  We could easily go back to the 5-line fix...

But the fundamental problem was that `*args, **kwargs` were being handled differently between  `figure.legend` and `axes.legend` for no good reason, and it will just be a maintenance headache in the future.  Worse the `legend.Legend` keywords were described three different places in three different formats, inconsistently in each place.  

The changes are actually pretty small.

  1. moved the argument-parsing logic out of `axes.legend` into `legend._parse_legend_args`
  2. used the same parsing in `figure.legend`
  3. changed the three docstrings to refer to `legend_kw_doc`. 

I need to do some cleanup, and I still have no idea how to address the failing `mock` test (see comment above), but otherwise I hope this is an improvement.

jklymak commented:
Oh, hmmmm.  `fig.legend()` did this funky thing where if you called 

```python
fig, axs = plt.subplots(2,1)
l1 = axs[0].plot(x, 'r', label='boo')
l2 = axs[1].plot(x, 'r', label='boo')
fig.legend()
```
It would only put 'boo' in the legend once.  This PR doesn't respect this.  Easy to fix, but do we really want to?

anntzer commented:
I guess we "have to" for backcompatibilty but that sounds like yet another thing I'd like to get rid of... (for example Axes.legend does not provide the same functionality).

jklymak commented:
@anntzer   Ahem, now it does ;-)

anntzer commented:
I can't say I really hate this feature (so if others like it I'm fine with it), but in my view it's a bit similar to auto reuse of axes (https://github.com/matplotlib/matplotlib/pull/9037 and linked isses) and would prefer not having it for the same reason.

jklymak commented on lib/matplotlib/legend.py:
```diff
@@ -973,3 +1110,136 @@ def draggable(self, state=None, use_blit=False, update="loc"):
+    handles = []
+    labels = []
+
+    def _in_handles(h, l):
```
@anntzer  This is where the logic to remove duplicate label/line pairs is.  Easy to remove if there is a consensus.

tacaswell commented on lib/matplotlib/figure.py:
```diff
@@ -1462,67 +1351,16 @@ def legend(self, *args, **kwargs):
-                            '(artist handles, figure labels, legend location)')
-
-        l = Legend(self, handles, labels, **kwargs)
+        print("Axes:", self.axes)
```
rouge print

tacaswell commented on lib/matplotlib/legend.py:
```diff
@@ -973,3 +1110,136 @@ def draggable(self, state=None, use_blit=False, update="loc"):
+
+    # Two arguments:
+    #   * user defined handles and labels
+    elif len(args) >= 2:
```
This should be `==`, passing `loc` as a positional argument in deprecated for the Axes version but apparently we missed doing the same to the Figure version (as it did not explicitly take `loc` as a positional, but passed `*args` through to the `Legend.__init__` so it worked.

This should probably return `handles, labels, extra_args` and in `Axes.legend` raise if `len(extra_args) > 0` and pass them through in `Figure.legend`

jklymak replied:
OK, this is caught in `lib/matplotlib/axes/_axes.py`.    But it still needs to be in here.

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -510,57 +328,13 @@ def legend(self, *args, **kwargs):
+        kwargs.pop('handles', None)
+        kwargs.pop('labels', None)
+        # check for third arg
+        if len(args) == 3:
```
This was intentionally deprecated in 1.5 http://matplotlib.org/api/api_changes.html#legend it should not come back!

tacaswell commented:
Modulo the handling of a positional `loc` being inconsistent between axes and figure :+1:

jklymak commented:
@tacaswell  Can we just deprecate the third positional argument in both?  Right now the code doesn't care if its called from `figure` or `axes`.

tacaswell commented:
In the spirit of minimal change / semver I would rather deprecate `loc` on `Figure.legend` in 2.2 rather than 2.1.1.  It is already _removed_ from `Axes.legend` from 1.5 so we don't need to deprecate it there ;)

jklymak commented:
@tacaswell  Allowing the third positional argument is back in `Axes.legend` with this PR, because *all* the code to handle the arguments is the now same between `Axes.legend` and `figure.legend`.  Thats how this PR got rather large-looking.  If we handle `figure` and `Axes` differently, their code bases will diverge again, and thats how we ended up in this mess in the first place!

I'm suggesting we silently go back to allowing the third positional argument in `Axes.legend`, but deprecate if anyone happens to use it, and deprecate it in `figure.legend` as well.  Then we can take it out again for 2.2.  Its not like it hurts anything to handle the third positional argument.

jklymak commented:
Ah OK, I understand.  Thats easy enough, if a bit of a PITA.  

Not sure I understand your last comment, though.  Currently only `handles`, `labels`, and `loc` can be passed as positional, and `loc` actually gets converted to a `kwarg`.

tacaswell commented:
In 2.0 the truncated `Figure.legend` is

```python

def legend(self, handles, labels, *args, **kwargs):
    l = Legend(self, handles, labels, *args, **kwargs)
    self.legends.append(l)
    l._remove_method = lambda h: self.legends.remove(h)
    self.stale = True
    return l
```

so the API change from 2.0 -> 2.1 is bigger than just breaking `handlers` and `labels` as kwargs.  If we are going to go back to the < 2.1 behavior we should bring it all back, add deprecation warnings to positional args in 2.2 and then remove in the next release.

tacaswell commented:
Sorry, I was using confusing tense!

jklymak commented:
Did anyone complain at 2.0 when the change was made?   

If you look at the 1.5.3 doc string, the kwarg positions and the documented positions aren't even the same:

```python
__init__(self, parent, handles, labels,
                 loc=None,
                 numpoints=None,    # the number of points in the legend line
                 markerscale=None,  # the relative size of legend markers
                                    # vs. original
```

```python
"""
Optional keyword arguments:
        ================   ====================================================
        Keyword            Description
        ================   ====================================================
        loc                a location code
        prop               the font property
        fontsize           the font size (used only if prop is not specified)
"""
```

My wild guess is that no-one was using the ability to pass `numpoints` as a positional argument...

jklymak commented:
Docs build locally, so not sure why there is a circle/ci error

jklymak commented:
Rebased and squashed previous commits.  

@tacaswell I think this is ready to go.  However, it does *not* roll back to allowing more than three positional arguments.  If you really think it should, I'm happy to work on it, but I am having trouble seeing that anyone would have done that.  See comment above...

QuLogic commented:
I'm inclined to say that a large refactoring should go into 2.2, and just the bugfix should go into a point release unless this is difficult to do nicely.

jklymak commented:
@QuLogic  My biggest concern is that rebasing this work after a few months is probably not going to be trivial.  

The change is actually quite small, aside from the refactoring which makes it look big.

tacaswell commented:
@jklymak That change was something that was missed review when the original PR went in,  I would fix all of the API change rather than just the part someone flagged (users are very clever and _will_ find and use surprising parts of the API).

@QuLogic I'm not super worried about the size of this patch for the micro-release

@jklymak This is against master so we could just merge this as-is, do a simpler change on 2.1.x to fix the 2 API breaks and then handle the merge conflict carefully.

jklymak commented:
@tacaswell   I'm not 100% sure what you'd like.

If anyone used positional arguments past `loc`, then they used them inconsistently with the `1.5.3` documentation.  The 1.5.3 docs for [`figure.legend`](https://github.com/matplotlib/matplotlib/blob/v1.5.3/lib/matplotlib/figure.py) (line 1209) say the next argument is `props`, but the actual next keyword in [legend.Legend](https://github.com/matplotlib/matplotlib/blob/v1.5.3/lib/matplotlib/legend.py) (line 151) was `numpoints`. So anyone who did a call like `fig.legend(handles, labels, loc, npoints)` was doing so by reading the source code.

The fix is trivial, so its not like I mind doing it.  I just don't think its warranted:

```python
if len(extra_args):
            cbook.warn_deprecated(
                "2.1",
                "Figure.legend will accept no more than two "
                "positional arguments in the future.  Use "
                "'fig.legend(handles, labels, loc=location)' "
                "instead.")
            kwargs['loc'] = extra_args[0]
            extra_args = extra_args[1:]
        l = mlegend.Legend(self, handles, labels, *extra_args, **kwargs)
```

NelleV commented on lib/matplotlib/figure.py:
```diff
@@ -1327,130 +1328,18 @@ def legend(self, *args, **kwargs):
-            Default is *None* which will take the value from the
-            ``legend.columnspacing`` :data:`rcParam<matplotlib.rcParams>`.
+
+        %(_legend_kw_doc)s
```
At some point, it had been decided to stop using docstring interpolation. They often break and are hard to maintain.

I don't know where that is.

jklymak replied:
That is unfortunate, but I'm more than happy to do this the "right" way.   

Is there a better way to not write out the same doc string three times?  The fundamental issue here was that `figure.legend`, `axes.legend` and `Legend.__init__` all have the same kwargs, and *all three* had a different version of the documentation for them.  So that is also hard to maintain, and was definitely broken.   There has to be a way to share parts of doc strings, and I assumed this was it based on the existing code base.  

As for discoverability, most IDEs will let you search for a string in a project.  `_legend_kw_doc` is in `legend.py`.  

However, if there is a better way to specify the doc string so its easy to maintain and share parts of the string, I'm more than happy to switch!   Thanks.

NelleV replied:
No, there isn't an easy way not to write the docstring three times, apart from docstring interpolation.

For very "common" options, the idea was to link back to one place containing the information (linestyle options, colors, markers, etc). 
In that particular case, I would just copy paste the kwargs.

The question is whether you want to favor consistency over simplicity. I favor simplicity, as I believe documentation should be easily editable by anyone. Some contributors believe that consistency is more important.

jklymak replied:
I think the interpolation string is editable by anyone, though I'll readily admit I had considerable trouble getting the indentation correct!   But I didn't find it particularly complicated.  

Looking at the code base, I think the historic problem is that people used names like `Axes` for the documentation string, and that is truly undiscoverable.  

I think consistency is pretty important.  If we add or subtract a kwarg from `legend` in the future, someone has to remember to change three doc strings, or have the documentation be wrong in the future.  The whole mess with this PR is that the docs were not consistent and the kwargs were not handled consistently, so it lead to user complaints.  

I'm happy to do as I'm told.  But I think having a way to include common docstrings is pretty powerful.  I'd certainly agree, however, that they should have a strong naming convention so they are easily searchable.  i.e. maybe `module_function_(*_)_docstring` (where the (*_) is an optional descriptor).

NelleV replied:
Experiences showed over and over again that our docstring hacking makes Matplotlib's documentation extremely hard to edit by beginners in contributing to Matplotlib, yet fairly advanced python users.

FYI, that includes core-developers not noticing that our html documentation was eating part of our docstrings for years.

jklymak commented:
Fair enough.  

I don’t like the option of referring to another method or class for the keywords.   If I’m in IPython or a notebook I want tab to give me the docs for the function I’m working on and referring to another class breaks that discoverability

So to me that leaves hard copying the docstring three times.  

If that is the way forward, then I wonder if there is a way to put a test that says the Parameter docstring for these three functions should be identical.  That way a developer will know they need to keep these consistent?  Not sure how such a test is written, but...

> On Oct 14, 2017, at 10:03 AM, Nelle Varoquaux <notifications@github.com> wrote:
> 
> @NelleV commented on this pull request.
> 
> In lib/matplotlib/figure.py:
> 
> > -            font-size units.
> -            Default is *None* which will take the value from the
> -            ``legend.handletextpad`` :data:`rcParam<matplotlib.rcParams>`.
> -
> -        borderaxespad : float or None
> -            The padding between the axes and legend border, measured in
> -            font-size units.
> -            Default is *None* which will take the value from the
> -            ``legend.borderaxespad`` :data:`rcParam<matplotlib.rcParams>`.
> -
> -        columnspacing : float or None
> -            The spacing between columns, measured in font-size units.
> -            Default is *None* which will take the value from the
> -            ``legend.columnspacing`` :data:`rcParam<matplotlib.rcParams>`.
> +
> +        %(_legend_kw_doc)s
> Experiences showed over and over again that our docstring hacking makes Matplotlib's documentation extremely hard to edit by beginners in contributing to Matplotlib, yet fairly advanced python users.
> 
> FYI, that includes core-developers not noticing that our html documentation was eating part of our docstrings for years.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jklymak commented:
@tacaswell  Checked in version now allows more than three positional arguments for `figure.legend`.  Also explicitly disallow more than two positional arguments for `axes.legend`.

Curious for the input of other devs on the docstring interpolation issue noted above in my conversation with @NelleV .  I poked around, and it seems very easy to include a test in `test_legend.py` to make sure the `Parameter` section is the same between all three dosctrings.  If docstring interpolation is indeed fragile, I can implement that instead.

anntzer commented:
I agree with @NelleV that string interpolation as it is done now is way too flaky, and left a lot of rendering issues uncaught.  If we really want to interpolate in parameter descriptions (a worth goal IMO) this should be done in a more structured manner, say with some machinery so that something like
```
@interpolate_docstring
def foo():
    """
    blah blah

    {other_function:parameters}
    """
```
"works" (probably based on some rst parsing and re-rendering).

jklymak commented:
It has to work in `ipython` too in my opinion.  Right now the help strings for colorbar for instance work well in ipython.  Referring to another method does not.  i.e. for `ax.text()`:

```
Other Parameters
----------------
**kwargs : `~matplotlib.text.Text` properties.
    Other miscellaneous text parameters.
```

is next to useless for me as a user.

jklymak commented:
Given that two of you vote for *not* using the interpolation I will copy the parts of the docstrings, and implement a test in `test_legend.py` that checks the docstrings for consistency.  That will stop the docstrings from being edited in isolation.

NelleV commented:
@jklymak I'd wait until an official decisions has been made. We've been having those for several years now, and a final decision on that matter needs to be taken. (I do think we did converge to remove docstring interpolation and stop hacking our docstring, but I may be wrong)

jklymak commented:
@NelleV Even if I cut and paste, its an improvement on the current situation and would allow this PR to proceed.  I'll then open an issue for discussion with this as a good example of a place where interpolation is helpful.

jklymak commented:
@NelleV @anntzer Actually one point of clarification (and please feel free to point me towards the original conversation).  Was it the simple string substitutions that were flaky or the automatic kwarg passthroughs (which I don't fully understand)?    I wonder if a happy middle ground is what I was doing - define a fixed string that describes the kwargs that is that is shared rather than passed through.

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -353,7 +329,7 @@ def legend(self, *args, **kwargs):
 
         bbox_to_anchor : :class:`matplotlib.transforms.BboxBase` instance \
-or tuple of floats
+        or tuple of floats
```
This should not be indented (otherwise the rendering in ipython gets messed up).

anntzer commented:
https://github.com/matplotlib/matplotlib/issues/7095, or http://matplotlib.org/2.0.0/api/_as_gen/matplotlib.axes.Axes.semilogy.html#matplotlib.axes.Axes.semilogy, are example of issues caused by uncontrolled interpolation.

I understand the "but I'm very careful" argument but again I think we should either use a proper ("structured") interpolation machinery or none at all.

tacaswell commented on lib/matplotlib/figure.py:
```diff
@@ -1462,67 +1518,22 @@ def legend(self, *args, **kwargs):
+        kwargs.pop('labels', None)
+        # check for third arg
+        if len(extra_args):
+            cbook.warn_deprecated(
```
It is too late to deprecate this for 2.1 if we take semver seriousishly.

jklymak replied:
Oops, sorry.  What should go there?  `2.1.1`?  `2.2`?

QuLogic replied:
What @tacaswell means is this PR can't be backported if the deprecation is made (or at least the deprecation part should be a separate PR that isn't backported.)

jklymak replied:
OK, so remove the deprecation for now, and then when its merged make a new PR?   Thats easy enough....

tacaswell replied:
Thanks, sorry for being pesky about this.

jklymak replied:
No problem at all.  I'm new here, and don't have any experience with the versioning in or outs.

QuLogic replied:
Deprecations are [point 7 in semver](http://semver.org/#spec-item-7); you can read the rest at your leisure.

tacaswell commented on lib/matplotlib/figure.py:
```diff
@@ -1462,67 +1518,22 @@ def legend(self, *args, **kwargs):
+        handles, labels, extra_args = mlegend._parse_legend_args(self.axes,
+                *args, **kwargs)
+        # these are popped in the above...
+        kwargs.pop('handles', None)
```
might be a bit cleaner to have `_parse_legend_args` also return `kwargs` so that you do not have to remember to pop these two?

jklymak replied:
I'll be honest, I couldn't figure out the right syntax because I only have a faint idea what `**` means in `**kwargs`!

jklymak commented:
@anntzer @NelleV  OK, sorry to be annoying about this, but I do wonder if we are throwing out the baby with the bathwater.  

Here, I just wanted to do a simple:

```python

boo_kw_doc = '''

    ============  ====================================================
    Property      Description
    ============  ====================================================
    *booit*      [ 'neither' | 'both' | 'min' | 'max' ]
                  If not 'neither', make pointed end(s) for out-of-
                  range values.  These are set for a given colormap
                  using the colormap set_under and set_over methods.
```

and then in some docstring:

```python

        Parameters
        ------------

        %(boo_kw_doc)s

       Notes
       ------

       Here is a note...
 
```

So, its just substituting a string.  All the funky keyword stuff, I agree, was pretty complicated.  But this is a pretty straightforward way to substitute a string into the docstring.  

Anyway, I just want to make sure folks thought *that* use of this machinery was really a problem.

anntzer commented:
@tacaswell wanna chime in on the docstrings?  I'm against interpolation (as it is done right now) but would be happy to leave the final word to the real boss, hehe :-)

tacaswell commented:
The bad formatting is definitely a real problem (the issue, as I understand it, boils down to identifying and managing indentation in the inserted strings) as is the opaqueness of how the system works.  @NelleV taught a summer school last summer (2 summers ago?  I am bad at time...) where our docs were the learning platform and there was a fair amount of difficulty.

On the other hand, in the terminal links are not great, on the other other hand, some of the tables we were putting in via the paramaterization were so long that they just made the docstring almost unusable.

On the 4th hand copy-pasted docstrings are pretty annoying (as they _will_ get out of sync, but the test is a very good idea!).

On the 5th hand, I am not sure how much our docstring handling slows us down at import time. 

Now that I have agreed and disagreed with everyone on at least one point :smiling_imp: :angel:, I think the best path forward here is the copy-paste docstrings and a long term plan to modernize the docstring interpolation code.

jklymak commented:
Squashed.  

This sounds fine to me.  I think encouraging a test when people find themselves cutting and pasting is a decent stopgap measure to stop documentation from drifting.  The little `get_docstring_section` function I wrote could be made more robust and put in `cbook`?  `docstring`?

jklymak commented:
ping!  I don't think there were any more changes requested, or if there is, I'm happy to work on it.

anntzer commented on lib/mpl_toolkits/axes_grid1/parasite_axes.py:
```diff
@@ -258,8 +258,8 @@ def get_aux_axes(self, tr, viewlim_mode="equal", axes_class=None):
 
-
     def _get_legend_handles(self, legend_handler_map=None):
+        # don't use this!
```
Is this intended for deprecation (then just delete it, it's private) or is it just used privately (then that's already implied by the underscore and the comment doesn't add anything)?

jklymak replied:
Fair enough - I was cautious because I don't really grok everything in `axes_grid1` and I didn't want just cut some code that someone may use...

jklymak replied:
rebased...

tacaswell replied:
@jklymak That is a good instinct!

anntzer commented:
If we are going to break backcompat either on Axes.legend or on Figure.legend wrt. duplicated entries then I would prefer *removing* the functionality from Figure.legend (with a proper note) rather than *adding* it to Axes.legend.

jklymak commented:
> If we are going to break backcompat either on Axes.legend or on Figure.legend wrt. duplicated entries then I would prefer removing the functionality from Figure.legend (with a proper note) rather than adding it to Axes.legend.

OK, right now we aren't breaking any backwards compatibility, though I did talk about dropping the positional kwargs. I re-did it so that all the old behaviour was respected, and didn't add any hidden behaviour.   Are you referring to the future?

anntzer commented:
Unless I am mistaken (tbh I haven't paid a lot of attention, so I may have), this changes the behavior of
```
ax.plot([0, 1], c="b", label="foo")
ax.plot([0, 1], c="b", label="foo")
ax.legend()
```
from having `foo` twice in the legend to having it once (matching the behavior for Figure).

jklymak commented:
Ah, only if they have exactly the same linestyles as well.

anntzer commented:
Probably none.
I will concede that duplicate suppression is actually more likely to be useful across multiple Axes, but the "keep the semantics simple" part of myself is cringing :)

jklymak commented:
I understand the concern, and I would never have written that code myself w/o prodding.    OTOH I think this is genuinely useful; you can imagine plotting the same quantity type more than once even in the same axes instance, and wanting it labeled as such in the plot call, but not want to have to worry about pruning the extra legend entries (or putting an underscore at the beginning of the label string, which is a pretty obscure UI).

dstansby commented on lib/matplotlib/figure.py:
```diff
@@ -1462,67 +1517,21 @@ def legend(self, *args, **kwargs):
+            #    "positional arguments in the future.  Use "
+            #    "'fig.legend(handles, labels, loc=location)' "
+            #    "instead.")
+            kwargs['loc'] = extra_args[0]
```
What happens here if `'loc'` is also already in kwargs?

jklymak replied:
Its dropped!  I suppose we could do something here, but given that we don't want to encourage this third element, I'm not sure catering to its misuse is really necessary...

dstansby commented on lib/matplotlib/legend.py:
```diff
@@ -281,6 +403,15 @@ def __init__(self, parent, handles, labels,
             setattr(self, name, value)
         del locals_view
 
+        # trim handles and labels if illegal label...
```
👍

dstansby commented on lib/matplotlib/legend.py:
```diff
@@ -974,3 +1105,145 @@ def draggable(self, state=None, use_blit=False, update="loc"):
+
+    ``ax.legend()`` is equivalent to ::
+
+      h, l = ax.get_legend_handles_labels()
```
Should this be `ax._get_legend_handles_labels()` (missing underscrore)?

jklymak replied:
Not really - there is a `ax.get_legend_handles_labels()`  But it doesn't need to be referenced here!

dstansby commented on lib/matplotlib/tests/test_legend.py:
```diff
@@ -14,6 +14,36 @@
+
+
+# test that docstrigs are the same
+def get_docstring_section(func, section):
```
👍 I like this a lot!

jklymak replied:
Yes, its saved me already, though the cutting and pasting is getting old...

jklymak commented:
Believe it or not, `ax.legend(arg1)` can *also* accept just a list of handles, a fact I found out only because @choldgraf did it in http://matplotlib.org/devdocs/gallery/text_labels_and_annotations/custom_legends.html#sphx-glr-gallery-text-labels-and-annotations-custom-legends-py

*But*, there is no way to attach a proper label to the legend when you call the legend like this, even if the underlying handles have labels!  In master:

```python 

N = 10
data = [np.logspace(0, 1, 100) + np.random.randn(100) + ii for ii in range(N)]
data = np.array(data).T
cmap = plt.cm.coolwarm
rcParams['axes.prop_cycle'] = cycler(color=cmap(np.linspace(0, 1, N)))

fig, ax = plt.subplots()
lines=[]
for i in range(N):
    lines += [ax.plot(data[:, i], label='%s'%i)]
    print(lines[-1])
ax.legend(lines)
```

<img width="579" alt="figure_1" src="https://user-images.githubusercontent.com/1562854/31911214-c4febff0-b7f4-11e7-87c1-875c1a49a3cc.png">

That is broken beyond belief in my opinion, because:

```python 
fig, ax = plt.subplots()
lines=[]
for i in range(N):
    lines += [ax.plot(data[:, i], label='%s'%i)]
    print(lines[-1])
ax.legend(lines[::3])
```
is a completely reasonable thing to do, but yields:

<img width="587" alt="figure_1" src="https://user-images.githubusercontent.com/1562854/31911289-04cc5e76-b7f5-11e7-81e1-962cb956453a.png">

I won't fix in this PR, but wanted to register my dismay.  

I think `ax.legend(arg)` should accept either a list fo strings (old label format), or a list of handles.  If the first, it does what it does now and just finds the first N handles to attach the strings to.  If the latter, it provides a legend for that list.  

The arg = list of handles is not documented, and I think @choldgraf example above should be modified because it is doing something `legend` isn't really meant to do (all tests and docs passed for my change before this example was put into master).  I think we should add support for lists of handles, but only display them if the label is explicitly set, and document this as *new* behaviour.  If you'd like that as a 2.2 separate PR, thats fine.  I think the PR here, as it stands, gives us back what we had before 2.1.  Pinging @tacaswell @anntzer @dstansby Thanks!

anntzer commented:
As your second example shows, `ax.legend` is not "accepting a list of handles" (you will note that the artists that went into the legend in the second example are not every third, but the first 4); instead, it is interpreting the argument as a list of labels (after forcefully converting them using `str`) and assigning them as labels to the first *n* artists.

Yes this is nonsensical, but should really be the object of a separate PR.

jklymak commented:
This is ready to go, unless we find something *else* whacky that the various versions of legend do...

I have two approvals, but I assume its bad form to self-merge?

tacaswell commented:
@jklymak I think we did something funny with git :(

jklymak commented:
Sorry!  My fault.  Let me re-push.  I made a local change and was pushing onto your change.  But I got your change.  I won't touch it until tonight, honest!

jklymak commented:
@dopplershift Yes fair enough - its easy to fix the docs later.  I'll add a note that this is one place to fix in #9414 if its not there already....

jklymak commented on lib/matplotlib/figure.py:
```diff
@@ -1462,67 +1518,22 @@ def legend(self, *args, **kwargs):
+            #     "instead.")
+            # kwargs['loc'] = extra_args[0]
+            # extra_args = extra_args[1:]
+            pass
```
@tacaswell  This now fails Travis - what was the plan here?  I thought we were keeping `figure.legend` w/ three possible args?

tacaswell replied:
This shouldn't have changed anything other than to prevent `fig.legend(a, b, l, loc=c)` case.

jklymak replied:
Ah, I see:

```
self = <matplotlib.tests.test_legend.TestLegendFigureFunction object at 0x7fffd6b1fc50>
    def test_legend_label_three_args(self):
        fig, ax = plt.subplots()
        lines = ax.plot(range(10))
        with mock.patch('matplotlib.legend.Legend') as Legend:
            fig.legend(lines, ['foobar'], 'right')
>       Legend.assert_called_with(fig, lines, ['foobar'], loc='right')
```

The test was checking that `loc` got loaded w/ the third element.  I'll fix later. Thanks!

jklymak replied:
@tacaswell  Changed `legend_test` to check for an Exception should someone call: `fig.legend(handles, labels, 'right', loc='left')`  (`test_legend_label_three_args_pluskw`)  I think thats what you wanted.  

Any reason to leave this check/pass in here?  Just posterity sake?

tacaswell replied:
To avoid changing the API as much as possible.  I am :+1: on deprecating the other positional arguments for 2.2, but not in 2.1.1 (so leave that code in place so we can just un-comment it in a follow on PR).

I don't think we needed a test for what is core-python behavior (you can not pass a value as both positional and keyword) nor should we break it!

jklymak commented:
Ummm, help!    The original example no longer works!!

```python 

fig, ax = plt.subplots(1, 2)
hs = []
lab = ['Boo', 'Hoo']
for nn, axx in enumerate(ax):
    h, = axx.plot(np.arange(10))
    hs += [h]
fig.legend(hs, lab)
plt.show()

```

```

/Users/jklymak/matplotlib/lib/matplotlib/legend.py:932: UserWarning: Legend does not support [<matplotlib.lines.Line2D object at 0x107cfc3c8>] instances.
A proxy artist may be used instead.
See: http://matplotlib.org/users/legend_guide.html#creating-artists-specifically-for-adding-to-the-legend-aka-proxy-artists
  "aka-proxy-artists".format(orig_handle)
/Users/jklymak/matplotlib/lib/matplotlib/legend.py:932: UserWarning: Legend does not support [<matplotlib.lines.Line2D object at 0x107cfc908>] instances.
A proxy artist may be used instead.
See: http://matplotlib.org/users/legend_guide.html#creating-artists-specifically-for-adding-to-the-legend-aka-proxy-artists
  "aka-proxy-artists".format(orig_handle)
```

jklymak commented:
Ooops, false alarm.  I had a typo in my test!

```python

import numpy as np
import matplotlib.pyplot as plt

fig, ax = plt.subplots(1, 2)
hs = []
for axx in ax:
    h, = axx.plot(np.arange(10))
    hs += [h]
fig.legend(handles=hs, labels=['Boo', 'Hoo'])
plt.show()
```
 works fine!
